Added a filter to sort categories by position by default

Dnyomo opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
This fixed a bug where the default GraphQL Category method does not sort by category position. By adding this filter results are returned in the correct order.

### Related Pull Requests
magento/catalog-storefront#173

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/catalog-storefront#104

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

Commit:
Added a filter to sort categories by position by default

Dnyomo commented:
@magento run WebAPI Tests

Commit:
Test Fixes for the new filter that sorts categories by position

Dnyomo commented:
@magento run WebAPI Tests

mslabko commented on app/code/Magento/CatalogGraphQl/Model/Category/CategoryFilter.php:
```diff
@@ -71,6 +72,7 @@ public function getResult(array $criteria, StoreInterface $store)
         $categoryIds = [];
         $criteria[Filter::ARGUMENT_NAME] = $this->formatMatchFilters($criteria['filters'], $store);
         $criteria[Filter::ARGUMENT_NAME][CategoryInterface::KEY_IS_ACTIVE] = ['eq' => 1];
+        $criteria[Sort::ARGUMENT_NAME][CategoryInterface::KEY_POSITION] = ['ASC'];
```
@Dnyomo can you please verify and confirm that for categories (e.g in menu) also have a sort order by position ASC ?

Commit:
Added a new default sort by ID when searching Elasticsearch in SearchCriteriaBuilder.php
Remove the _id from the order that is being passed to the product collection in ProductSearch.php
Updated the position to be DESC passed to the product collection in ProductSearch.php
Updated the ProductSearchTest.php to have the items in the attended order

Commit:
Added a new default sort by ID when searching Elasticsearch in SearchCriteriaBuilder.php
Remove the _id from the order that is being passed to the product collection in ProductSearch.php
Updated the position to be DESC passed to the product collection in ProductSearch.php
Updated the ProductSearchTest.php to have the items in the attended order

Dnyomo commented:
@magento run WebAPI Tests, Integration Tests

mslabko commented on app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/ProductSearch.php:
```diff
@@ -145,15 +149,25 @@ private function getSearchResultsApplier(
         if (is_array($sortOrders)) {
             foreach ($sortOrders as $sortOrder) {
-                $ordersArray[$sortOrder->getField()] = $sortOrder->getDirection();
+                if ($sortOrder->getField() !== '_id') {
```
Why do we need this check? It will ignore any sort order defined by user in GQL query

mslabko commented on app/code/Magento/CatalogGraphQl/DataProvider/Product/SearchCriteriaBuilder.php:
```diff
@@ -104,6 +104,7 @@ public function build(array $args, bool $includeAggregation): SearchCriteriaInte
             $this->addDefaultSortOrder($searchCriteria, $isSearch);
         }
 
+        $this->addEntityIdSort($searchCriteria, $isSearch);
```
could you please verify the behavior for GQL query which will filter by category id/ids, e.g.
```
{
  products(filter: {category_id: {in: ["3", "5"]}}) {
   items {    
      sku
    }
  }
}
 
{
  products(filter: {category_id: {eq:"3"}}) {
   items {    
      sku
    }
  }
}
 ```

In these examples we should add order by "position ASC" only for the second case. 
And only if the sort was provided explicitly in GQL query it should be added as is.

I'm afraid that current implementation in addDefaultSortOrder method don't take this into account 

I think we dont' have specific tests for that and it make sense to add this for GQL coverage
(@yaroslavGoncharuk can you help us with steps?)

Dnyomo replied:
I did test this and because this method (\Magento\Catalog\Model\ResourceModel\Product\Collection::addAttributeToSort line 1749) converts the attribute position to entity_id. I think it makes sense to always have a position set as a sort.

Commit:
Updated tests to cover changes to the sort orders

Dnyomo commented:
@magento run WebAPI Tests

Commit:
Remove the position from the filter and replaced it with Entity Id

Dnyomo commented:
@magento run WebAPI Tests

Commit:
Removed bad code that was not needed

Dnyomo commented:
@magento run WebAPI Tests

yaroslavGoncharuk commented:
@Dnyomo Query with condition **eq** returns "Internal server error"
```
{
  products(filter: {category_id: {eq:"3"}}) {
   items {    
      sku
    }
  }
} 
```
Works well if **eq** is replaced with **in** OR **sort: {price: ASC}** argument is added

Commit:
Updated the default sort to check if the category id is passed as an array.
Reverted the tests back to check on products

Dnyomo commented:
@magento run WebAPI Tests

Commit:
Updated the default sort to check if the category id is passed as an array.
Reverted the tests back to check on products

Dnyomo commented:
@magento run WebAPI Tests

Commit:
Updated test

Dnyomo commented:
@magento run WebAPI Tests

Commit:
missed this test

Dnyomo commented:
@magento run WebAPI Tests

Dnyomo commented:
@magento run Integration Tests

Dnyomo commented:
> @Dnyomo Query with condition **eq** returns "Internal server error"
> 
> ```
> {
>   products(filter: {category_id: {eq:"3"}}) {
>    items {    
>       sku
>     }
>   }
> } 
> ```
> 
> Works well if **eq** is replaced by **in** OR **sort: {price: ASC}** argument is added

This issue has been fixed and all Integration and WebAPI Tests have passed.

mslabko committed:
Merge branch '2.4-develop' into storefront-issue_104_default_sort_order_b

mslabko commented:
@magento run all tests

mslabko commented:
@magento run Functional Tests

mslabko commented:
@Dnyomo  can you please check static tests?
https://github.com/magento/magento2/runs/941354354

mslabko commented:
@magento run Functional-Tests-CE tests

mslabko commented:
@magento run Functional Tests

mslabko commented:
@magento run Functional Tests CE,Functional Tests B2B

mslabko commented:
@magento run Functional Tests B2B

Commit:
Code style fixes

Commit:
Merge branch 'storefront-issue_104_default_sort_order_b' of github.com:Dnyomo/magento2 into storefront-issue_104_default_sort_order_b

Dnyomo commented:
@magento run Static Tests

Commit:
Code style fixes

Dnyomo commented:
@magento run Static Tests

Commit:
Code style fixes and added a message around converting _id to entity_id

Dnyomo commented:
@magento run Static Tests

Commit:
Updated the message with the related method

Dnyomo commented:
@magento run Static Tests

Commit:
Remvoed @SuppressWarnings(PHPMD.CouplingBetweenObjects)

Dnyomo commented:
@magento run Static Tests

mslabko committed:
Merge branch '2.4-develop' into storefront-issue_104_default_sort_order_b

mslabko commented:
@magento run all tests

mslabko commented:
@magento run Functional Tests B2B
